File: leetcode/range-sum-of-bst.kt
// Runtime: 348 ms, faster than 90.48% of Kotlin online submissions for Range Sum of BST.
// Memory Usage: 47.2 MB, less than 25.00% of Kotlin online submissions for Range Sum of BST.

/**
* Example:
* var ti = TreeNode(5)
* var v = ti.`val`
* Definition for a binary tree node.
* class TreeNode(var `val`: Int) {
*     var left: TreeNode? = null
*     var right: TreeNode? = null
* }
*/
class Solution {
  fun rangeSumBST(root: TreeNode?, L: Int, R: Int): Int {
    if (root == null) {
      return 0;
    }

    val v = root.`val`;

    if (v > R) {
      return rangeSumBST(root.left, L, R);
    } else if (v < L) {
      return rangeSumBST(root.right, L, R);
    } else {
      return rangeSumBST(root.left, L, R) + rangeSumBST(root.right, L, R) + v;
    }
  }
}

File: leetcode/range-sum-of-bst.js
// Runtime: 188 ms, faster than 61.20% of JavaScript online submissions for Range Sum of BST.
//   Memory Usage: 67.1 MB, less than 40.00% of JavaScript online submissions for Range Sum of BST.

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} L
 * @param {number} R
 * @return {number}
 */
var rangeSumBST = function(root, L, R) {
  if (root === null) {
    return 0;
  }

  const { val, left, right } = root;

  const lSum = val >= L ? rangeSumBST(left, L, R) : 0;
  const rSum = val <= R ? rangeSumBST(right, L, R) : 0;
  const selfVal = val >= L && val <= R ? val : 0;

  return lSum + rSum + selfVal;
};
